tests/kani/MemPredicates/foreign_type.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Z mem-predicates -Z c-ffi
+//! Check that Kani's memory predicates return that it's not safe to access pointers with foreign
+//! types since it cannot compute its size.
+#![feature(ptr_metadata)]
+#![feature(extern_types)]
+
+extern crate kani;
+
+use kani::mem::{can_dereference, can_read_unaligned, can_write};
+use std::ffi::c_void;
+use std::ptr;
+
+#[derive(Clone, Copy, kani::Arbitrary)]
+struct Wrapper<T: ?Sized, U> {
+    _size: U,
+    _value: T,
+}
+
+extern "C" {
+    type Foreign;
+    type __CPROVER_size_t;
+    fn __CPROVER_havoc_object(p: *mut c_void);
+
+}
+
+#[kani::proof]
+pub fn check_write_is_unsafe() {
+    let mut var: Wrapper<u64, usize> = kani::any();
+    let ptr: *mut Wrapper<Foreign, usize> = &mut var as *mut _ as *mut _;
+    assert!(!can_write(ptr));
+}
+
+#[kani::proof]
+pub fn check_read_is_unsafe() {
+    let var: usize = kani::any();
+    let ptr = &var as *const _ as *const __CPROVER_size_t;
+    assert!(!can_dereference(ptr));
+    assert!(!can_read_unaligned(ptr));
+}
+
+/// Kani APIs cannot tell if that's safe to write to a foreign type.
+///
+/// However, foreign APIs that have knowledge of the type can still safely set new values.
+#[kani::proof]
+#[kani::should_panic]
```

Comment:
I don't understand this test. If we're trying to see if the foreign API can safely set new values, doesn't it defeat the point to make the harness panic? Shouldn't we do the write and see that it succeeds, and then perhaps do the failing read in a separate test?

Reply:
I can split this into two tests as you suggested. My initial thought was just to check that the value has changed by checking if it is still 0.